### PR TITLE
Add Draft RFC8 & 9 (ERC20)

### PR DESCRIPTION
**Pull request by LLFourn**
This is branched off #38 (which is branched off #35).

I haven't read over them yet but it might be ready for review.

### CODE REVIEW

File: RFC-007-SWAP-Ethereum-basic.md [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+| Ledger   | Identity Name | JSON Encoding         | Description         |
+|:---------|:--------------|:----------------------|---------------------|
+| Ethereum | `address`     | `0x` prefixed address | An Ethereum Address |
+
+# Examples
+
+## RFC003 SWAP REQUEST
+
+This following shows an [RFC003](RFC-003-SWAP-basic.md) SWAP REQUEST where the `alpha_ledger` is Ethereum, the `alpha_asset` is 1 Ether (with `...` being used where the value is only relevant for the `beta_ledger`).
```
Comment (bonomat) [inline RFC-007-SWAP-Ethereum-basic.md]:
```suggestion
The following shows an [RFC003](RFC-003-SWAP-basic.md) SWAP REQUEST where the `alpha_ledger` is Ethereum, the `alpha_asset` is 1 Ether (with `...` being used where the value is only relevant for the `beta_ledger`).
```

File: RFC-009-SWAP-ERC20-basic.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+To refund the HTLC, the funder MUST send a transaction to `contract_address` with empty data in a block with timestamp greater than `expiry`.
+
+To be notified of the refund event, both parties SHOULD watch the blockchain for  `contract_address` emitting the `Refunded()` log.
+
+# Examples
+
+## RFC003 SWAP REQUEST
+
+This following shows an [RFC003](RFC-003-SWAP-basic.md) SWAP REQUEST where the `alpha_ledger` is Ethereum, the `alpha_asset` is 1 PAY token (with `...` being used where the value is only relevant for the `beta_ledger`).
```
Comment (bonomat) [inline RFC-009-SWAP-ERC20-basic.md]:
```suggestion
The following shows an [RFC003](RFC-003-SWAP-basic.md) SWAP REQUEST where the `alpha_ledger` is Ethereum, the `alpha_asset` is 1 PAY token (with `...` being used where the value is only relevant for the `beta_ledger`).
```

File: RFC-005-SWAP-Bitcoin-basic.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+| Ledger  | Identity Name | JSON Encoding            | Description                                                                           |
+|:--------|:--------------|:-------------------------|---------------------------------------------------------------------------------------|
+| Bitcoin | `pubkeyhash`  | `hex-encoded-bytes (20)` | The result of applying SHA256 and then RIPEMD160 to a SECP256k1 compressed public key |
+
+# Examples
+
+## RFC003 SWAP REQUEST
+
+This following shows an [RFC003](RFC-003-SWAP-basic.md) SWAP REQUEST where the `alpha_ledger` is Bitcoin, the `alpha_asset` is 1 Bitcoin (with `...` being used where the value is only relevant for the `beta_ledger`).
```
Comment (bonomat) [inline RFC-005-SWAP-Bitcoin-basic.md]:
```suggestion
The following shows an [RFC003](RFC-003-SWAP-basic.md) SWAP REQUEST where the `alpha_ledger` is Bitcoin, the `alpha_asset` is 1 Bitcoin (with `...` being used where the value is only relevant for the `beta_ledger`).
```

File: RFC-007-SWAP-Ethereum-basic.md [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    "beta_asset": {...},
+    "protocol": {
+        "value" : "comit-rfc-003",
+        "parameters" : { "hash_function" : "SHA-256" }
+    }
+  },
+  "body": {
+    "aplha_ledger_refund_identity": "0x0f59e9e105be01d5e2206792a267406f255c5ea5",
+    "alpha_expiry": 1552263040,
+    "secret_hash" : ""adebd583a094215e963ebe4a1474b9bb4bf48167e64f3f474d71e41a75494bbb"",
```
Comment (bonomat) [inline RFC-007-SWAP-Ethereum-basic.md]: The double quotes were probably not on purpose:
```suggestion
    "secret_hash" : "adebd583a094215e963ebe4a1474b9bb4bf48167e64f3f474d71e41a75494bbb",
```

File: registry.md [hunk 5]
```diff
@@ -7,35 +27,88 @@ This registry may be expanded with new RFCs.
...
+| `ethereum` | [RFC-006](./RFC-006-SWAP-Ethereum.md)| The Ethereum family of blockchains     |
+
+
+And the possible parameters they each may have:
+
+### `bitcoin` Parameters
+
+| Parameter | Value Type                                  | Description                              |
+|:----------|---------------------------------------------|------------------------------------------|
+| `network` | see [Bitcoin Networks](#bitcoin-networks) | The particular blockchain netowrk to use |
```
Comment (bonomat) [inline registry.md]:
```suggestion
| `network` | see [Bitcoin Networks](#bitcoin-networks) | The particular blockchain network to use |
```

File: registry.md [hunk 6]
```diff
@@ -7,35 +27,88 @@ This registry may be expanded with new RFCs.
...
 
-| Ledger   | Identity Name | JSON Encoding            | Reference | Description                                                                       |
-|:----     |:-------      |:-------------            |:--------- | --------------------------------------------------------------------------------- |
-| Bitcoin  | `pubkeyhash`  | `hex-encoded-bytes (20)` | TBD       | The result of applying SHA256 and then RIPEMD160 to a user's SECP256k1 public key |
-| Ethereum | `address`     | Ethereum address        | TBD       | An Ethereum address                                        |
+### `ethereum` Parameters
 
+| Parameter | Value Type                                    | Description                              |
+|:----------|-----------------------------------------------|------------------------------------------|
+| `network` | see [Ethereum Networks](#ethereum-networks) | The particular blockchain netowrk to use |
```
Comment (bonomat) [inline registry.md]:
```suggestion
| `network` | see [Ethereum Networks](#ethereum-networks) | The particular blockchain network to use |
```

File: registry.md [hunk 7]
```diff
@@ -7,35 +27,88 @@ This registry may be expanded with new RFCs.
...
+| `address`  | `0x` prefixed address | The address of the ERC20 contract                                           |
+
+
+## Identities
+
+[RFC003](./RFC-003-SWAP-basic.md#identity) requires that each ledger has an associated identity:
+
+| Ledger   | Identity Name | JSON Encoding            | Reference                            | Description                                                                           |
+|:---------|:--------------|:-------------------------|:-------------------------------------|---------------------------------------------------------------------------------------|
+| Bitcoin  | `pubkeyhash`  | `hex-encoded-bytes (20)` | [RFC-004](./RFC-004-SWAP-Bitcoin.md) | The result of applying SHA256 and then RIPEMD160 to a SECP256k1 compressed public key |
```
Comment (bonomat) [inline registry.md]:
```suggestion
| Bitcoin  | `pubkeyhash`  | `hex-encoded-bytes (20)` | [RFC-004](./RFC-004-SWAP-Bitcoin.md) | The result of applying SHA-256 and then RIPEMD-160 to a SECP256k1 compressed public key |
```

File: RFC-005-SWAP-Bitcoin-basic.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+At the start of the deployment stage, both parties compile the contract into bytes as described in the previous section.
+We will call this value `contract_script`.
+
+To deploy the Bitcoin HTLC, the *funder* must confirm a transaction on the relevant Bitcoin blockchain.
+One of the transaction's outputs must have the following properties:
+
+- Its `value` MUST be equal to the `quantity` parameter in the Bitcoin asset header.
+- It MUST have a Pay-To-Witness-Script-Hash (P2WSH) `scriptPubKey` derived from `contract_script` (See [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#specification) for how to construct the `scriptPubkey` from the `contract_script`).
+
+The redeeming party (the *redeemer*) should wait until an transaction with the above output is included in the Bitcoin blockchain with enough confirmations such that they consider it permanent.
```
[the remark this answers is not in the document]
Comment (LLFourn) [reply RFC-005-SWAP-Bitcoin-basic.md]: This is already fixed in the bitcoin branch.

File: RFC-006-SWAP-Ethereum.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+| `network` | see [Ethereum Networks](#ethereum-networks) | The particular blockchain network to use |
+
+
+#### Ethereum Networks
+
+And adds a Ethereum Networks section describing the possible values for this parameter:
+
+| Value     | Description                      | Network Id |
+|:----------|:---------------------------------|------------|
+| `regtest` | Private Ethereum regtest network | N/A        |
```
[the remark this answers is not in the document]
Comment (LLFourn) [reply RFC-006-SWAP-Ethereum.md]: I will move this comment to the Ethereum PR and respond to it there.

File: RFC-007-SWAP-Ethereum-basic.md [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+
+| Topic        | keccack256(Topic)                                                  | Data     | Emitted When             |
+|:-------------|:-------------------------------------------------------------------|----------|:-------------------------|
+| `Redeemed()` | 0xb8cac300e37f03ad332e581dea21b2f0b84eaaadc184a295fef71e81f44a7413 | `secret` | The contract is redeemed |
+| `Refunded()` | 0x5d26862916391bf49478b2f5103b0720a842b45ef145a268f2cd1fb2aed55178 | -        | The contract is refunded |
+
+
+Implementations SHOULD use the following gas limits on the transactions related to the contract:
+
+| Transaction | recommended gas limit |
```
[the remark this answers is not in the document]
Comment (LLFourn) [reply RFC-007-SWAP-Ethereum-basic.md]: It's what we use in the code. (or at least what I calculated from it).

After specifying this we should just use these values directly.

File: RFC-005-SWAP-Bitcoin-basic.md [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+This may be expanded in subsequent RFCs.
+
+### Parameters
+
+The parameters for the Bitcoin HTLC follow [RFC003](./RFC-003-SWAP-basic.md) and are described concretely in the following table:
+
+| Variable        | Description                                                                 |
+|:----------------|:----------------------------------------------------------------------------|
+| asset           | The quantity of satoshi                                                     |
+| secret_hash     | The hash of the `secret` (32 bytes for SHA-256)                             |
```
[the remark this answers is not in the document]
Comment (LLFourn) [reply RFC-005-SWAP-Bitcoin-basic.md]: Hmm I wouldn't know where to reference. The secret is defined in the HTLC section of RFC003. I could make the above reference exactly that section. I will comment this on the Bitcoin PR (and the others which have the same sentence).

File: RFC-009-SWAP-ERC20-basic.md [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+  "body": {
+    "aplha_ledger_refund_identity": "0x0f59e9e105be01d5e2206792a267406f255c5ea5",
+    "alpha_expiry": 1552263040,
+    "secret_hash" : "adebd583a094215e963ebe4a1474b9bb4bf48167e64f3f474d71e41a75494bbb",
+    "beta_ledger_redeem_identity" : "...",
+    "beta_expiry" : ...
+  },
+}
+```
+
```
Comment (LLFourn) [inline RFC-009-SWAP-ERC20-basic.md]: ❌ Include the pre-image of the `secret_hash`.

File: RFC-008-ERC20.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+To describe ERC20 asset in an Asset type header specify `erc20` as the value along with the following parameters:
+
+### `quantity`
+
+The `quantity` paramter describes the quantity of token wei the asset represents.
+The `quantity` parameter is mandatory.
+Its value MUST be a `u256` (note that in the JSON encoding a `u256` is encoded as decimal string like `"1000000000000000000"`).
+
+### `address`
```
Comment (thomaseizinger) [inline RFC-008-ERC20.md]: :x: I preferred the parameter name `token_contract`. Maybe we could also go for `token_identity`?

Comment (thomaseizinger) [reply RFC-008-ERC20.md]: `address` sounds so generic ...

File: RFC-009-SWAP-ERC20-basic.md [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+They SHOULD do this by sending a contract deployment transaction to the relevant Ethereum blockchain with `contract_code` as the `data` of the transaction.
+
+
+To be notified of the deployment event, both parties MAY watch the blockchain for a transaction with the `contract_code` as the data.
+Upon observing the deployment transaction, both parties SHOULD record the address the contract was deployed to (referred to as `contract_address` from now on).
+
+
+### Funding
+
+After deploying the contract, the funder must transfer ownership of the `token_quantity` to the `contract_address`.
```
Comment (thomaseizinger) [inline RFC-009-SWAP-ERC20-basic.md]: :x: `contract_address` is now the 3rd way of referring to the token contract :)

I think this should be more consistent, especially when you are using backticks to emphasize the way it is spelled.

Comment (LLFourn) [reply RFC-009-SWAP-ERC20-basic.md]: contract_address is the HTLC contract address. But yeah let's make this explicit.

File: RFC-008-ERC20.md (hunk 13 above)
Comment (LLFourn) [reply RFC-008-ERC20.md]: Agreed. This was just me specifying how it's already implemented.

**Pull request merged**